### PR TITLE
Add file upload support for chat sessions (#19)

### LINKED ISSUE #19: Add support for file uploading
### Description

Add support for uploading files to the LLM during a chat session. This will allow users to provide better context or additional data (e.g., logs, images) that the model can reference when generating responses. 

### Upstream changes

_No response_

### Are you interested in contributing this feature?

_No response_

### PR DESCRIPTION
**Pull request by GunaPalanivel**
## Summary
Adds the ability for users to attach files when sending messages to the chatbot, giving the LLM richer context for troubleshooting Jenkins issues.

### Backend (Python/FastAPI)
- New `file_service.py` module for file processing and validation
- Added multipart form upload endpoint `POST /sessions/{session_id}/message/upload`
- Added `GET /files/supported-extensions` endpoint for frontend validation
- Updated `chat_service.py` to incorporate file content into LLM prompts
- New Pydantic models: `FileAttachment`, `ChatRequestWithFiles`, `SupportedExtensionsResponse`

### Frontend (React/TypeScript)
- Added attach button (📎) to the input component
- File preview chips showing name, size, and remove option
- Image thumbnails and file icons displayed in chat messages
- Client-side validation for file size and extension

### Supported File Types
- **Logs/Text:** `.txt`, `.log`, `.md`
- **Code:** `.py`, `.js`, `.ts`, `.tsx`, `.java`, `.groovy`, `.sh`
- **Config:** `.json`, `.xml`, `.yaml`, `.yml`, `.properties`, `.toml`
- **Jenkins:** `Jenkinsfile`, `.groovy`
- **Images:** `.png`, `.jpg`, `.jpeg`, `.gif`, `.webp`, `.bmp`

### Limits
- Max file size: 10MB per file
- Images are base64 encoded for LLM vision capabilities

## Testing
- Added unit tests for file service (`test_file_service.py`)
- Added route tests for upload endpoint (`test_file_upload.py`)
- Updated frontend tests for Input component

## How to Test
1. Start the backend and frontend servers
2. Click the 📎 button or drag files into the input area
3. Attach a log file or screenshot and ask a question
4. Verify the file content appears in the chat and the LLM references it

Closes #19

### CODE REVIEW

File: chatbot-core/api/services/chat_service.py [hunk 1]
```diff
@@ -28,34 +29,57 @@
...
+    # Include file info in memory message
+    user_message = user_input
+    if files:
+        file_names = [f.filename for f in files]
+        user_message = f"{user_input}\n[Attached files: {', '.join(file_names)}]"
+    
+    memory.chat_memory.add_user_message(user_message)
     memory.chat_memory.add_ai_message(reply)
 
     return ChatResponse(reply=reply)
```
Comment (cnu1812) [inline chatbot-core/api/services/chat_service.py]: Overall suggestions:
- We need to ensure upload_file.close() is called for each file in a finally block to prevent temporary file accumulation on the server.
- The default message 'Please analyze the attached file(s).' is hardcoded. Should we move this to a constant in config.py or the service layer so it's easier to tune the system prompt later?
- In the Exception catch-all, using str(e) can sometimes leak internal path information or sensitive library errors. For production-ready code, we might want to log the full error and return a more generic message to the user.
- Is process_uploaded_file a heavy operation (e.g., OCR or large text parsing)? If so, and if it's a synchronous function, we should be careful about blocking the FastAPI event loop.

File: chatbot-core/tests/unit/routes/test_file_upload.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    ]
+    
+    response = client.post(
+        "/sessions/test-session-id/message/upload",
+        data={"message": ""},
+        files=files
+    )
+    
+    assert response.status_code == 200
+
```
Comment (cnu1812) [inline chatbot-core/tests/unit/routes/test_file_upload.py]: Verify that the "Truncation" logic in file_service.py works when called through the API. by creating `def test_chatbot_reply_text_truncation(client, mock_session_exists, mock_get_chatbot_reply):`

File: chatbot-core/tests/unit/services/test_file_service.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        assert is_text_file("Jenkinsfile") is True
+        assert is_text_file("Dockerfile") is True
+        assert is_text_file("Makefile") is True
+
+    def test_rejects_non_text_files(self):
+        """Test that non-text files are rejected."""
+        assert is_text_file("image.png") is False
+        assert is_text_file("photo.jpg") is False
+        assert is_text_file("archive.zip") is False
+
```
Comment (cnu1812) [inline chatbot-core/tests/unit/services/test_file_service.py]: Jenkins and Docker environments often use hidden files or files without extensions (e.g., .env, .gitignore). so handle this edge case.

File: chatbot-core/api/services/chat_service.py (hunk 1 above)
Comment (GunaPalanivel) [reply chatbot-core/api/services/chat_service.py]: **Resolution:**

1. Added `finally: await upload_file.close()` in the route handler
2. Kept the default message in the route layer (can be moved to config if needed)
3. Changed error message to use `type(e).__name__` instead of `str(e)` to avoid leaking internal details:

```python
except Exception as e:
    raise HTTPException(
        status_code=500,
        detail=f"Failed to process file: {type(e).__name__}"
    ) from e
finally:
    await upload_file.close()
```

File: chatbot-core/tests/unit/services/test_file_service.py (hunk 3 above)
Comment (GunaPalanivel) [reply chatbot-core/tests/unit/services/test_file_service.py]: **Resolution:** Updated `is_text_file()` to recognize hidden files and added tests:

```python
def test_recognizes_hidden_files(self):
    """Test that hidden files (dotfiles) are recognized as text files."""
    assert is_text_file(".env") is True
    assert is_text_file(".gitignore") is True
    assert is_text_file(".dockerignore") is True
    assert is_text_file(".editorconfig") is True
    assert is_text_file(".eslintrc") is True
    assert is_text_file(".prettierrc") is True
```

File: frontend/src/tests/Input.test.tsx [hunk 4]
```diff
@@ -50,17 +57,17 @@ describe("Input Component", () => {
...
     );
-    expect(screen.getByRole("button")).toBeDisabled();
+    expect(screen.getByRole("button", { name: /send/i })).toBeDisabled();
 
     rerender(
       <Input input="    " setInput={mockSetInput} onSend={mockOnSend} />,
     );
-    expect(screen.getByRole("button")).toBeDisabled();
+    expect(screen.getByRole("button", { name: /send/i })).toBeDisabled();
   });
```
[the remark this answers is not in the document]
Comment (GunaPalanivel) [reply frontend/src/tests/Input.test.tsx]: **Resolution:** Added test:

```tsx
it("keeps send button disabled with whitespace-only input and no files", () => {
  render(<Input input="   " ... attachedFiles={[]} />);

  expect(screen.getByRole("button", { name: /send/i })).toBeDisabled();
});
```

File: chatbot-core/api/services/file_service.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+    Args:
+        content: The file content as bytes.
+        filename: The name of the file.
+
+    Returns:
+        Tuple[str, str]: A tuple of (base64_encoded_content, mime_type).
+    """
+    mime_type, _ = mimetypes.guess_type(filename)
+    if mime_type is None:
```
[the remark this answers is not in the document]
Comment (berviantoleo) [reply chatbot-core/api/services/file_service.py]: Based on the current reasoning. I don't think it makes sense for an image file.

If mime_type can't produce a valid image mimetype, we should consider it's not a valid image file.

File: chatbot-core/api/services/file_service.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        ext = get_file_extension(filename)
+        mime_map = {
+            ".png": "image/png",
+            ".jpg": "image/jpeg",
+            ".jpeg": "image/jpeg",
+            ".gif": "image/gif",
+            ".webp": "image/webp",
+            ".bmp": "image/bmp"
+        }
+        mime_type = mime_map.get(ext, "application/octet-stream")
```
Comment (berviantoleo) [inline chatbot-core/api/services/file_service.py]: This might be a dangerous thing to do. We should ensure the returned mimetype is a valid image file. Instead of trying to fallback to searching the mimetype by ourselves when the mimetypes can't do.

File: chatbot-core/api/services/file_service.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    Args:
+        filename: The name of the file.
+
+    Returns:
+        str: The lowercase file extension including the dot (e.g., ".txt").
+    """
+    return Path(filename).suffix.lower()
+
+
+def is_text_file(filename: str) -> bool:
```
[the remark this answers is not in the document]
Comment (GunaPalanivel) [reply chatbot-core/api/services/file_service.py]: You're absolutely right - this is a legitimate security consideration. Relying solely on file extensions is not robust since someone could rename `malware.exe` → `malware.txt`.

I'm implementing magic byte validation now to address this:

1. Adding `python-magic` library for MIME type detection from actual file content
2. For images: verify magic bytes match the claimed extension
3. For text files: detect if content is actually an executable disguised as text

The validation runs before any file processing occurs. Will commit once it's done!

File: chatbot-core/api/services/file_service.py (hunk 6 above)
Comment (GunaPalanivel) [reply chatbot-core/api/services/file_service.py]: Good catch! I'm addressing this along with the other security concern.

The fix implements `validate_file_content_type()` that checks magic bytes **before** we reach this fallback. So the flow becomes:

1. Check file size
2. Validate content matches extension (magic bytes)
3. Only then process the file

If magic bytes don't match the extension, we reject immediately - this fallback will never be reached for invalid content.

Working on it now, will commit when done!
